Foundation/src/ActiveThreadPool.cpp

```diff
@@ -15,347 +15,493 @@
 	{
-		ActiveThread* pThread = createThread();
-		_threads.push_back(pThread);
-		pThread->start();
+		_pool->noActiveThreads.broadcast();
 	}
 }
 
 
-ActiveThreadPool::ActiveThreadPool(std::string  name, int capacity, int stackSize):
-	_name(std::move(name)),
-	_capacity(capacity),
-	_serial(0),
-	_stackSize(stackSize),
-	_lastThreadIndex(0)
+ActiveThreadPoolPrivate::ActiveThreadPoolPrivate(int capacity, int stackSize_):
+	maxThreadCount(capacity),
+	stackSize(stackSize_)
 {
-	poco_assert (_capacity >= 1);
+}
 
-	_threads.reserve(_capacity);
 
-	for (int i = 0; i < _capacity; i++)
-	{
-		ActiveThread* pThread = createThread();
-		_threads.push_back(pThread);
-		pThread->start();
-	}
+ActiveThreadPoolPrivate::ActiveThreadPoolPrivate(int capacity, int stackSize_, const std::string& name_):
+	name(name_),
+	maxThreadCount(capacity),
+	stackSize(stackSize_)
+{
 }
 
 
-ActiveThreadPool::~ActiveThreadPool()
+ActiveThreadPoolPrivate::~ActiveThreadPoolPrivate()
+{
+	joinAll();
+}
+
+
+bool ActiveThreadPoolPrivate::tryStart(Runnable& target)
 {
-	try
+	if (allThreads.empty())
 	{
-		stopAll();
+		startThread(target);
+		return true;
 	}
-	catch (...)
+
+	if (activeThreadCount() >= maxThreadCount)
 	{
-		poco_unexpected();
+		return false;
 	}
-}
 
+	if (!waitingThreads.empty())
+	{
+		// recycle an available thread
+		enqueueTask(target);
+		ActivePooledThread* pThread = waitingThreads.front();
+		waitingThreads.pop_front();
+		pThread->notifyRunnableReady();
+		return true;
+	}
 
-int ActiveThreadPool::capacity() const
-{
-	return _capacity;
+	if (!expiredThreads.empty())
+	{
+		// restart an expired thread
+		ActivePooledThread* pThread = expiredThreads.front();
+		expiredThreads.pop_front();
+
+		++activeThreads;
+
+		// an expired thread must call join() before restart it, or it will cost thread leak
+		pThread->join();
+		pThread->setRunnable(target);
+		pThread->start();
+		return true;
+	}
+
+	// start a new thread
+	startThread(target);
+	return true;
 }
 
 
-void ActiveThreadPool::start(Runnable& target)
+inline bool comparePriority(int priority, const QueuePage* p)
 {
-	getThread()->start(Thread::PRIO_NORMAL, target);
+	return p->priority() < priority;
 }
 
 
-void ActiveThreadPool::start(Runnable& target, const std::string& name)
+void ActiveThreadPoolPrivate::enqueueTask(Runnable& target, int priority)
 {
-	getThread()->start(Thread::PRIO_NORMAL, target, name);
+	for (QueuePage* page : runnables)
+	{
+		if (page->priority() == priority && !page->isFull())
+		{
+			page->push(target);
+			return;
+		}
+	}
+	auto it = std::upper_bound(runnables.begin(), runnables.end(), priority, comparePriority);
+	runnables.insert(it, new QueuePage(target, priority));
 }
 
 
-void ActiveThreadPool::startWithPriority(Thread::Priority priority, Runnable& target)
+int ActiveThreadPoolPrivate::activeThreadCount() const
 {
-	getThread()->start(priority, target);
+	return (int)(allThreads.size() - expiredThreads.size() - waitingThreads.size());
```

Comment:
Change return type to std::size_t to avoid type casting or use C++ instead of C casting.